tests/test_single_topology_v3.py

```diff
@@ -0,0 +1,180 @@
+        all_idxs, _ = setup_dummy_interactions_from_ff(
+            ff, mol, dummy_group=[12], root_anchor_atom=6, nbor_anchor_atom=4
+        )
+
+
+def test_find_dummy_groups_and_anchors():
+    """
+    Test that we can find the anchors and dummy groups when there's a single core anchor atom. When core bond
+    is broken, we should disable one of the angle atoms.
+    """
+    mol_a = Chem.MolFromSmiles("OCCC")
+    mol_b = Chem.MolFromSmiles("CCCF")
+    core = np.array([[1, 2], [2, 1], [3, 0]])
+
+    dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core[:, 0], core[:, 1])
+    assert dgs == [{3}]
+    assert jks == [(2, 1)]
+
+    # angle should swap
+    core = np.array([[1, 2], [2, 0], [3, 1]])
+
+    with pytest.warns(CoreBondChangeWarning):
+        dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core[:, 0], core[:, 1])
+        assert dgs == [{3}]
+        assert jks == [(2, None)]
+
+
+def testing_find_dummy_groups_and_multiple_anchors():
+    """
+    Test that we can find anchors and dummy groups with multiple anchors, we expect to find only a single
+    root anchor and neighbor core atom pair.
+    """
+    mol_a = Chem.MolFromSmiles("OCC")
+    mol_b = Chem.MolFromSmiles("O1CC1")
+    core = np.array([[1, 1], [2, 2]])
+
+    with pytest.warns(MultipleAnchorWarning):
+        dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core[:, 0], core[:, 1])
+        assert dgs == [{0}]
+        assert jks == [(1, 2)] or jks == [(2, 1)]
+
+    mol_a = Chem.MolFromSmiles("C(C)(C)C")
+    mol_b = Chem.MolFromSmiles("O1CCCC1")
+
+    core_a = [0, 1, 2, 3]
+    core_b = [2, 1, 4, 3]
+
+    with pytest.warns(MultipleAnchorWarning):
+        dgs, jks = single_topology_v3.find_dummy_groups_and_anchors(mol_a, mol_b, core_a, core_b)
+        assert dgs == [{0}]
+        assert jks == [(1, 2)]
+
+
+def compute_distance_matrix(x0):
+    N = x0.shape[0]
+    D = x0.shape[1]
+    xi = x0.reshape((1, N, D))
+    xj = x0.reshape((N, 1, D))
+    rij = xi - xj
+    dij = jnp.linalg.norm(rij, axis=-1)
+    return dij
+
+
+def get_max_distance(x0):
+    dij = compute_distance_matrix(x0)
+    return jnp.amax(dij)
+
+
+def test_hif2a_end_state_stability():
+    """
+    Pick some random pairs from the hif2a set and ensure that they're numerically stable at the
+    end-states under some a standard distance based atom_mapping protocol.
+    """
+    suppl = Chem.SDMolSupplier("timemachine/testsystems/data/ligands_40.sdf", removeHs=False)
+    mols = [m for m in suppl]
+    pairs = []
+    for mol_a in mols:
+        for mol_b in mols:
+            pairs.append((mol_a, mol_b))
+
+    np.random.seed(2023)
+    np.random.shuffle(pairs)
+    ff = Forcefield.load_from_file("smirnoff_1_1_0_sc.py")
+
+    batch_distance_check = jax.vmap(get_max_distance)
+
+    # this has been tested for up to 50 random pairs
+    for mol_a, mol_b in pairs[:5]:
```

Comment:
Nice! This randomized stress-testing is good.

* Should we add a version of this test with a larger number of pairs to nightly tests?
* Is there a cheaper condition that can be tested for a large number of pairs? (E.g. omitting `simulate_system`, and just checking that system can be created & energy is finite on initial combined conf?)

Reply:
1)  yes but there are other things i'd also like to add to the nightly
2) good idea, however, the MCS code is also surprisingly slow for some of these pairs, not just the MD. But the MCS could be cached, and we could probably verify end-states can have their energies be proper minimized for a larger set?

Reply:
3. jit itself can be pretty slow (4-5 seconds per pair)